### PR TITLE
Alt Council Title (Custom Localisation) - Ready

### PR DESCRIPTION
**Pull request by Charanraja123**
First Commit.
As you can see this only requires two files, which are both our own and not the game file.
I overrode some game keys to call custom localization within the localisation file .I just went through the dev post where it said its enough for you to have the keys (wherever in the localization folder) and not the whole file, so used that method
So, One off topic thing here i want to say ,I think we can also rename cultures localization file and it should work.

### CODE REVIEW

**Comment by Charanraja123** [comment]
Do not Commit to the main file, if you have any suggestions or questions let me know. The problem with custom localisation structure is that its not as stable as flavorisation. A single mistake can cause the whole thing to fail (from experience).
Obviously you can still edit the localisation file like any other one.

**Comment by Gr770** [comment]
I mentioned this on the other branch but you still have to edit council_l_english.yml directly to get it to show up. The customizable localization can still be in a separate file.

I got it working by doing that.

**Comment by Gr770** [comment]
Also GetPlayer means that it will only check for the gender of the player not the councilor.

**Comment by Charanraja123** [comment]
@Gr770 Thats what I'm saying, in Ck3, unlike ck2, you can just copy the keys you want instead of editing the file. 

> 
> 
> Also GetPlayer means that it will only check for the gender of the player not the councilor.

I'll change this. My bad.

**Comment by Charanraja123** [comment]
> 
> 
> I mentioned this on the other branch but you still have to edit council_l_english.yml directly to get it to show up. The customizable localization can still be in a separate file.
> 
> I got it working by doing that.
> 
These files worked as is, without any alteration from my end. I didnt even have a council_l_english.yml in my mod. Please check again

**Comment by Gr770** [comment]
> > I mentioned this on the other branch but you still have to edit council_l_english.yml directly to get it to show up. The customizable localization can still be in a separate file.
> > I got it working by doing that.
> 
> These files worked as is, without any alteration from my end. I didnt even have a council_l_english.yml in my mod. Please check again

For whatever reason it is not working.

**Comment by Charanraja123** [comment]
> 
> 
> > > I mentioned this on the other branch but you still have to edit council_l_english.yml directly to get it to show up. The customizable localization can still be in a separate file.
> > > I got it working by doing that.
> > 
> > 
> > These files worked as is, without any alteration from my end. I didnt even have a council_l_english.yml in my mod. Please check again
> 
> For whatever reason it is not working.

Nope. You are right. I made some mistakes. Sorry

**Comment by Viridianus** [comment]
Are you sure it's a good practice to leave gender unspecified if the table currently has no female? (I mean, it's a possibility to be sure, but then the question arises why don't we do this for other types of titles, too...)

**Comment by Charanraja123** [comment]
The reason I took that step is because, if we restrict it to male only, we barely have any entries. If we get enough female entries in there, I could change it, if required. I have coded it such that I only check for males when there is a female version present. Otherwise I do not check gender.

**Comment by Charanraja123** [comment]
@Viridianus Did u just edit the game files?. You do realize that most of them are Vanilla game files (used for compatibility purposes) and are getting reworked by me?

Also, do not touch religion (even if it makes no sense), because we have not discussed it yet. Deleting localisation folder achieves absolutely nothing that you wish for. The players will now just get something like "councillor_court_chaplain_orthodox_county" in their screen. Which will not look good for our mod. 
As much as I understand and appreciate your enthusiasm, please be careful editing stuff. As I said when i created the pull request, Unless you are absolutely sure of what you are doing, do not edit these files. These are very fickle compared to flavorisations. I can(have) literally make all council positions in the game have a blank name through one missing bracket. Therefore i suggest leaving the editing to me or @Gr770 , especially for this file. You can still check our own localisation file where everything from the sheet is entered into.
I mean no offense, Next time just leave a comment and I'll be glad to sort it out for you.
 I apologize if I have offended you in any way .

**Comment by Viridianus** [comment]
I did not touch the big code file (and wrong bracketing tarnishing the whole code is expected behavior, I don't know why you keep stressing it). As for Orthodox - OK, there's an easy fix for that if they're _that_ bad with their code (implemented).

**Comment by Charanraja123** [comment]
This is also similar to the big code file as in it has an considerable impact on the game. Look I'm sorry if I came across Harsh, but you outright deleting those lines really threw me off. I don't mind you making edits or correcting things, but please be careful.

In any case, maybe you already knew this, `localization/english/ctp_council_custom_loc_l_english.yml` is where our defined localization are stored. In case you want to check them out too.

**Comment by Charanraja123** [comment]
@Viridianus  I would like to apologise to you. I just realised how assholeish i sounded, What you did was not even that big of deal, it was a simple mistake anyone could have done. I was just having a bad day on other terms, and took it out here. I'm extremely sorry, Hopefully we can still be friends xP

**Comment by Viridianus** [comment]
@Charanraja123 I can be even more harsh at times) We're okay)

**Comment by Charanraja123** [comment]
I've updated Steward Titles. xD

**Comment by Gr770** [comment]
Added Marshals and Coptic. See if that works for you, especially the coptic additions.

**Comment by Charanraja123** [comment]
Looks Good. 
Are you working on Spymaster or do you want me to add them in?

**Comment by Gr770** [comment]
I am not, I've been tooling with marches/palatinates.

**Comment by Charanraja123** [comment]
Okay. I think I'll wrap this up with Spymaster by today (my time, IST).

**Comment by Gr770** [comment]
I'm pretty sure I missed a bracket in Marshal somewhere, I'll check tomorrow.

**Comment by Charanraja123** [comment]
I'll anyway check, Once I've  done spy master stuff. Shouldn't be a problem.

**Comment by Charanraja123** [comment]
Finished Spy master. Check and let me know.

**Comment by Charanraja123** [comment]
I think I fixed all the errors, There are still some errors being generated in the log which I could not locate, I'm looking into it

**Comment by Charanraja123** [comment]
Actually they were some changes in the council vanilla file that we missed. I added them in. Next time we should be more careful

**Comment by Charanraja123** [comment]
SO.. I think I may made a big error.
The Spouse council coding is all within the Council Titles file. The one I fought so hard not to edit. I'm really sorry.
The Spouse localization given in our folder seems to only extend Holy Order, Mercenaries and Theocracies, I thought these were overall defaults. I still believe Spouse localization can be done with that single sentence, but we have edit that file. If you guys are okay with it, I'll edit in that too. 

Other than that the rest of them are completed and tested. 
Court Chaplain is only to be done after we enough data on theocracies. 

Again I'm sorry. I have no idea how I missed that.

**Comment by Viridianus** [comment]
Whoa, whoa, whoa, wait, wait, wait. Can't you create a customizable localization which calls for our flavorizations and then call the custom loc from localization, circumventing Council Titles file?

**Comment by Charanraja123** [comment]
Yes.. But, The council Titles file has to be coded in a way that exits come to the loc folder. For the spouse this is not the case unfortunately, It directly calls in the vanilla flavorisation from within the Council folder

Also. The custom loc that calls our flavorisation is the same thing `Character.GetTitleAsName`. and I wanted to call that from the localization.

In a sense you just suggested the same thing I suggested in the Issue (That was closed) but in a slightly more roundabout manner.

**Comment by Viridianus** [comment]
I have an idea for spouses - when you implement everything else, I'll try to code it (though it would be _much_ easier if our naming conventions were taking consorts not regnants as default, see post in #48).

Unrelated: you seem to use different naming convention for councillors, never calling on the cultural group's name but always listing cultures (ilmenian_severian_volhynian, bosnian_serbian, ...). Any particular reason - _other_ than not using cultural groups in the code of triggers?

Unrelated to both but **important**: our Vlach contributor of #63 has also proposed changes for Vlach councillors, and I "think" they are to be implemented.

**Comment by Charanraja123** [comment]
1. Everything is done from my side, except for spouses and court chaplain , so go for it.
2. Because , my personal reason is that unlike the flavourisation system, there is no priority here. If something qualifies for two conditions, the game just picks one at random. Therefore I wanted the name to reflect the unique conditions as much as possible. 
3. I have no idea why "think" is in quotes there lol. Yeah since there are already in the sheets, I can mod them in soon.

**Comment by Charanraja123** [comment]
Spouse Finished. 
@Viridianus Which one did you not understand, the whole thing or the special case?

**Comment by Viridianus** [comment]
The special case. Does "GetTitleAsName" only get the titles which are "special = holder"?

**Comment by Viridianus** [comment]
Also, please check your fork, I left a pull request there ;)

**Comment by Charanraja123** [comment]
@Viridianus Yes. It gets you Title based on your Landed status. 
Honestly, Even I was a little skeptical on whether It'll work for the special case.

**Comment by Charanraja123** [comment]
Actually, it makes complete sense now, That's why it check whether the Court Chaplain is landed before using `GetTitleAsName` for similar Council Titles

**Comment by Viridianus** [comment]
On the plus size, after all this coding I feel like I really understand how such code works)

As for multiple wives - shouldn't be a problem because Spouse position is automatically filled, and probably with the First Wife (haven't checked though). There's nothing in the code that would indicate you can switch the councilor wife if you have several.

**Comment by Charanraja123** [comment]
> On the plus size, after all this coding I feel like I really understand how such code works)

We all learn that way. I spent at least 30 hours exploring each and every file lol before I even came upon this mod xP

**Comment by Viridianus** [comment]
Though I feel like we should either replace the Pdx localizations for spouse or comment them out...

**Comment by Charanraja123** [comment]
I would not comment them out, because we have no idea where else those localisations are used. It can maybe be used for events and such.
I just to leave it be until someone reports a bug/error. I know it might sound bad, but we have no idea how and where that system works.

**Comment by Viridianus** [comment]
Then we could edit them to also call for title maybe?

**Comment by Charanraja123** [comment]
I mean, you can if you want to. 

Just remember that whatever you do there has zero chance of reflecting in the Council Screen and maybe has chances in being someone else but idk where. 

So I leave it to your choice.

**Comment by Viridianus** [comment]
Well, wherever it is called, it should be called properly not suddenly vanilized)

**Comment by Charanraja123** [comment]
Actually, One thing we can do is delete them, run the game and see if we get an error in the log. If no error, then it means those are never used. But I would need to play at least for hours to test that. 
Why don't we open a issue so that I remember it. Also because I don;t want wasting time editing things which might not even be used

**Comment by Viridianus** [comment]
Well, you certainly can open an issue, but I think that my fix should be sufficient. And as for whether they're called anywhere else - we can search for councillor_spouse in the files - if only 00_council_positions calls them, we're fine (for this patch at least).

**Comment by Charanraja123** [comment]
Yeah. No need for an Issue after that. I think we can actually merge this :D

**Comment by Viridianus** [comment]
Aha, so quick he is, "merge this we can". We've finished the _English_ version, my naïve friend. Now we'll have to look into French, German and Spanish.

**Comment by Viridianus** [comment]
Spanish seems to work.
German should work for titles which don't end in a sibilant (such as [s] or [ʃ]) - those (if male) should have a genitive in -es, not -s.
French possessives I can't make right as of now (they should be "du " before a male title which begins with a consonant, "de la " before female title which begins with a consonant, and "de l'" before any title which begins with a vowel).

**Comment by Viridianus** [comment]
Made a hack-fix for German (which will have to be updated each time we change a councilor title); I really don't think I have powers to do a similar thing for French to sort out the titles which begin with a vowel.

Also, this hack does not repair the "caution fix" - if somewhere `$councillor_spouse_king$` will shoot and give Kuningas, the possessive will be `des Kuningass` instead of expected `des Kuningases`. So we can only hope at least possessives will not shoot.

File: common/customizable_localization/ctp_DE_custom_loc_council.txt [hunk 1]
```diff
@@ -0,0 +1,81 @@
+DE_MaleCouncillorsHaveS = {
+	type = character
+	text = {
+		trigger = {
+			AND = {
```
Comment (Charanraja123) [inline common/customizable_localization/ctp_DE_custom_loc_council.txt:5]: Trigger by default is and, So this is redundant

**Review by Charanraja123 — COMMENTED** [review]
Ok. So, I just went through it only for a bit. But I already had some questions/suggestions

1) My first question would be whether you tested it?
2) These localization searches for the player culture while the council title is based on the liege culture. These might cause some weird stuff.
3) There are lots of ANDs and ORs (like the triggers plus 'AND' is redundant as I said before) which might cause errors. Try to minimize them if possible.

I could write up a draft for you, but if you would like to try out solving it yourself, you could do that too.

**Comment by Viridianus** [comment]
1. I have NOT tested it (otherwise I'd probably notice the liege thing).
2. OK, I'll change that.
3. I tried to get the logically correct version before thinking of simplifying it - that makes sense, right? Proceeding to delete the AND now.

File: common/customizable_localization/ctp_DE_custom_loc_council.txt (hunk 1 above)
Comment (Viridianus) [reply common/customizable_localization/ctp_DE_custom_loc_council.txt]: Changed.

**Comment by Viridianus** [comment]
@Charanraja123 - I can't use GitHub review request on you since it's your pull request, so I'll just call you like this)

**Comment by Charanraja123** [comment]
1. Well you should just test it then, if you feel its complete. Worst case, your game crashes
3. Well, Usually you want to test with something very simple and then build up on that. To get a better grip at things. 
It would be like
 a) First just test the council position localisation.
 b) then add culture and test again
 c) then add gender and test again.
And so on. The point being it always good to test it in a incremental manner to make sure your logic sticks.
Anyway, Coding practices aside, Do you want me to Test it and fix it? Or do you want to try yourself?

**Comment by Viridianus** [comment]
I want you to test it because my game takes ages to upload and double ages to add/modify a mod, not to mention it doesn't Alt+Tab very well.
As for coding - I kinda have some coding experience and I found that, well, my "blocks of understanding" are bigger than it's usually advised (if I test something too simple I won't have the generalizations I need).

**Comment by Charanraja123** [comment]
Okay. I'll test it.
I didn't mean to question your ability. I just listed out the process I follow, which I used for this Alt council too. You are free to follow what you are comfortable with.

**Comment by Viridianus** [comment]
Tell the results of your German test when you're finished - by the fact you're still not back I assume the game didn't crash on your head? One would actually also want to add spouse-clauses for duke and lower, but I guess it makes sense to test this one first.

**Comment by Charanraja123** [comment]
I haven't started yet, Was doing some other stuff

**Comment by Charanraja123** [comment]
There seems to be lot of encoding errors, Even for old files you edited.`ctp_council_custom_loc_l_german.yml` like this.
Make sure to UTF-8 BOM and not just UTF-8.
The game didn;t crash, but it deleted all title localization for some reason. I mean all of them. So I'm looking into that

**Comment by Charanraja123** [comment]
Okay.
@Gr770 and @Viridianus Have either of you tested/used any of the Non-English localization for Normal Titles. I'm not talking about council localization here. The Flavourisation Titles.

**Comment by Charanraja123** [comment]
image
I'm looking into the localization errors, But disregarding that your code doesn't seem to work, as show in the screenshot.
There some error with the independent command
```
Failed to find a valid event target link 'is_independent_ruler' at ' file: common/customizable_localization/ctp_DE_custom_loc_council.txt line: 89'
```
But even commenting that out it does not seem to work. But also no error pops up.

**Comment by Charanraja123** [comment]
I think its similar to what @Gr770 got for the Marches and Palatinates where it just didn;t work but also threw no errors

**Comment by Viridianus** [comment]
@Charanraja123 - recheck it after @Gr770's fix?

**Comment by Viridianus** [comment]
As for independent ruler: could something like `scope:liege = { is_independent_ruler = yes }` work instead?

**Comment by Gr770** [comment]
> As for independent ruler: could something like `scope:liege = { is_independent_ruler = yes }` work instead?

This is how it should be done. You can only use the chaining methods for scopes. so liege.culture = culture:whatever. is_independent_ruler is a trigger.


These are what you can use. https://ck3.paradoxwikis.com/Scopes#/media/File:CK3_Scopes.png

**Comment by Gr770** [comment]
> I'm looking into the localization errors, But disregarding that, your code doesn't seem to work, as show in the screenshot.
> But even commenting that out it does not seem to work. But also no error pops up.

Looks like it's having issues with DE_END_NOU_S_G and DE_ART_DEF_S_G not being defined.

**Comment by Charanraja123** [comment]
I think I have to ask the question of whether is this something we want to maintain? This is complex and has to be hard coded. It also changes every time we change something one of the Council Titles, Spouse Titles and in the future Court chaplains too. I'm also assuming a same thing will have to be done for Knights too. 

Maintaining all of this takes a lot of our time, Considering that we do not have much non English players (Going by the lack of any bug reports/comments on Steam). Is this something we want to waste time and resources on?

Please remember, this is not a one time thing. Someone will have to be doing this for years to come, it is a rather large commitment. I, for one, know that I'm not going to be as free as this in the future.

**Comment by Viridianus** [comment]
Updating this isn't going to be nearly as hard as creating it in the first place :D Also, worst-case scenario, we forget to update it and some German guy sees -shs instead of -shes. I doubt it's a cause for alarm.

> Looks like it's having issues with DE_END_NOU_S_G and DE_ART_DEF_S_G not being defined.

It's strange as they're vanilla, they're in 00_DE_custom_loc.txt.

**Comment by Viridianus** [comment]
> This is how it should be done.

Yet your fix omits the equals sign - intentional?

**Comment by Charanraja123** [comment]
Rebased the branch to get all those localisation fixes into here

**Comment by Charanraja123** [comment]
Okay.... So... council title possessive are never used. 
There's that. I'll search for any of the special titles if they were used. If not, we could just replace them with vanilla ones so at least they don't throw errors. And we could remove that custom localization for German.
I searched the entire game folder (not manually, using notepad ++), if you guys find anywhere it else let me know

**Comment by Charanraja123** [comment]
@Viridianus I think you can stop working on this, These are never used. :(

**Comment by Viridianus** [comment]
On French and German - go ahead. On English and Spanish - I think our decision is not harmful.

**Comment by Charanraja123** [comment]
We didn;t use any custom loc for possessive in english right. I dont mean the `$variable$` changes we made. That can stick. Just that one German one that you created. Rest can stay as it did.

**Comment by Viridianus** [comment]
Ah, sure. I have already commented that out so that file can go.

**Comment by Charanraja123** [comment]
Uhh.. I'll delete the file, cause its not being used? Do you want a copy of it or something?

**Comment by Viridianus** [comment]
> Do you want a copy of it or something?

I have one ;)

**Comment by Charanraja123** [comment]
Spanish - Tested and Working 
German - Tested and Working 
French - Tested and Working

**Pull request merged**